### PR TITLE
Super_context: add a cache for resolve_program

### PR DESCRIPTION
**Pull request by emillon**
On the test suite, the cache hit is 63%.

### CODE REVIEW

**Comment by emillon** [comment]
I extracted that from #1329 to make it easier to review.

File: src/super_context.ml [hunk 1]
```diff
@@ -598,6 +602,7 @@ let create
     ; file_tree
     ; stanzas_to_consider_for_install
     ; artifacts
+    ; artifacts_cache = Hashtbl.create 5
```
Comment (diml) [inline src/super_context.ml]: 5 feels small, I suggest to use 32

File: src/super_context.ml [hunk 2]
```diff
@@ -486,7 +487,10 @@ let dump_env t ~dir =
   Ocaml_flags.dump (Env.ocaml_flags t ~dir)
 
 let resolve_program t ?hint ~loc bin =
-  Artifacts.binary ?hint ~loc t.artifacts bin
+  Hashtbl.find_or_add
+    t.artifacts_cache
+    ~f:(Artifacts.binary ?hint ~loc t.artifacts)
```
Comment (rgrinberg) [inline src/super_context.ml]: I'm for the cache, but I think the loc and the hint should be part of the hash. With the current cache, we'll always be reporting the first loc encountered which isn't really correct.

Comment (diml) [reply src/super_context.ml]: Actually, I think it's better this way. For instance, if refmt is missing, there is no point reporting the error several times at different locations.

**Comment by emillon** [comment]
OK, thanks for double checking! Let's close this then.

**Pull request closed without merging**